Collect telemetry stats

linabutler opened the pull request:
This adds a simple recorder for timings and counts.

linabutler committed:
Collect timings and counters for telemetry.

linabutler committed:
Derive more `LogLevel` traits and clean up log macros.

thomcc commented on src/driver.rs:
```diff
@@ -58,6 +58,78 @@ pub enum LogLevel {
+/// Records timings and counters for telemetry.
+#[derive(Debug, Default)]
+pub struct Stats {
+    timings: HashMap<Timing, Duration>,
```
We know all possible values for both Timing and Counter, HashMap seems like overkill.

IMO you should either list them all out statically, or use a `Vec<Duration>` and make `enum Timing` and `enum Counter` be `#[repr(usize)]`.

Actually, now that I look, Counter already *is* listed out statically, as `StructureCounts`, so I think I am probably just not understanding the goal behind this PR.

linabutler replied:
Sorry, I should have explained where I was going with this. 😅

At first, I was thinking I could just pass `StructureCounts`, but there are other counts, like the local, remote, and merged tree size that aren't in there...though the merger could just report them, too, since it has them already, leaving us with just the timings. Then, I thought it'd be nice to have a wrapper for timings and counters that returned data in the format we want for Desktop's telemetry, like decimal seconds for times, so that's why I added `{Timing, Counter}Entry`.

I wanted to pass `Stats` to the merger and replace `StructureCounts`, but the merger can't hold a mutable reference while we also record stats in `Store::merge`, so that was a no-go. I also wanted to keep `StructureCounts` around for tests, so they didn't have to pass `Stats` everywhere. That's why `merge` copies values out of those counts and into `Stats`, which Desktop takes and passes to the merge callback.

Buuut...thinking more about it, I don't think we should record stats like in this PR, either. Rust merging on Desktop currently does it this way because I didn't have a way to have Rust on the storage thread to call multiple times into JS on the main thread. But I added that for logging, and I think it would make sense to do that for telemetry, too.

So, a better approach would be:

* Keep `StructureCounts`.
* Get rid of `Stats`.
* Add three methods to `Driver`: `record_structure_counts(&self, &StructureCounts)`, `record_timing(&self, Timing, Duration)`, and `record_counter(&self, Counter, i64)`.
* Get rid of duplicate `Counter`s that are already covered by `StructureCounts`.

On Desktop, these three methods would call into JS, which would call Sync's `recordTelemetryEvent`. So we'd record these events immediately, instead of accumulating them in a map and waiting until the end to report them.

linabutler replied:
WDYT?

thomcc replied:
Hmm, your call, although I probably would do it as returning the results rather than a callback-based approach.

The callback approach is more consistent with how things work on desktop already, but for the Rust-based merger it will probably end up having to jump through some (small) hoops to make that work for it's telemetry (at least it would need those to be `&mut self` or interior mutability to do so, but even that could get awkward easily to make the `&mut` viable there).

linabutler committed:
Replace `Stats` with a struct returned from `Store::merge`.

Since we already tracked most of our counts in `StructureCounts`, this
commit adds `merged_nodes` and `merged_deletions`. Bonus: counting
merged nodes this way lets us preallocate the space we need to hold all
the descendants when we recurse down the merged root.

This also adds a `MergedTimings` struct, and a macro for recording
timings with operation names.

linabutler commented:
Thanks, @thomcc! I like this much better. 😄

mhammond reviewed (approved):
lgtm
